### ISSUE TITLE
Tuple type mismatch with non-default-sized types

### ISSUE THREAD

**Issue by SLx64** [chapel-lang/chapel#6000]
### Summary of Problem

Tuples can't be assigned with non-default-sized types, e.g. int(8/16/32) or real(32).
 
### Steps to Reproduce

**Source Code:**

```chapel
var t: 3*int(32) = (1,1,1);
```

**Compile command:**
chpl -o test test.chpl

**Execution command:**
./test

### Configuration Information

- Output of `chpl --version`: chpl Version 1.15.0
- Back-end compiler and version: gcc (SUSE Linux) 4.8.1 revision 202388

**Comment by bradcray** [comment]
I agree that this is something we'd like to support, but will give an explanation for why we don't currently, as well as the challenges that we need to solve to get there:

First, as background, the reason that one can do assignments from int(64)s to int(32)s in other contexts like this one:

````
var i32: int(32) = 1;
````

Is because '1' is a param (compiler-known value), and the compiler knows that '1' can be stored as an int(32) without loss of information.  The assignment above would not be legal if the RHS was max(int(32))+1 since the value, though still a param, would not safely fit in the int(32).

This is what's known internally as param-coercion (the compiler will permit a coercion that it wouldn't normally allow because the value is known and known to be safe).  Note that if the int(64) is not a param, a cast is required:

````
var i64 = read(int(64));
var i32: int(32) = i64;                     // not permitted because we don't know that i64 will fit in i32
var i32b: int(32) = i64: int(32);           // OK because we're requesting an explicit downcast
var i32c: int(32) = i64.safeCast(int(32));  // OK if i64 fits in i32 at execution-time
````

There are several cases where we'd like to similarly support param coercions, such as your tuple-of-t case.  Others include range-of-idxType-t, domains-of-idxType-t, arrays-of-eltType-t, etc.  However because we don't support param tuples, ranges, domains, and arrays today, these cases don't fit into our current param-coercion scheme.

In thinking about what it would take to implement these more complex param cases, we have a difficult choice to make:  Do we explicitly teach the compiler how to reason about each of these types (tuples, ranges, domains, arrays) at compile-time?  Or do we teach the compiler how to reason about arbitrary records and classes at compile-time (which are used under the covers to implement most of these types) in which case any object can be a param and these just fall out?  The tradeoff is arguably that the former would be simpler but introduce more special case code (and duplicated code with the modules, arguably) while the latter is a much larger endeavor to take on.

We're also discussing the degree to which the param machinery might be handled by a Chapel interpreter/interactive development environment/REPL in the future -- i.e., if we could evaluate arbitrary Chapel code in a REPL, could that same code be used to evaluate arbitrary param expressions at compile-time?

**Comment by bradcray** [comment]
Note that this issue also relates somewhat to issues #5922 and #5343.

**Comment by lydia-duncan** [comment]
I added four future tests of this in #6003 .  @bradcray, how does the .future file look to you?

**Comment by bradcray** [comment]
I'd make them into feature request futures rather than bug futures.  I think the current implementation works as we'd expect, just not how we'd ultimately like.

**Comment by lydia-duncan** [comment]
Done!
